### PR TITLE
Predicate check array type

### LINKED ISSUE #4305: BE crash when array type predicate
### Steps to reproduce the behavior (Required)
```sql
CREATE TABLE `t1` (
  `c_1_0` decimal64(2, 1) NOT NULL COMMENT "",
  `c_1_1` datetime NOT NULL COMMENT "",
  `c_1_2` boolean NOT NULL COMMENT "",
  `c_1_3` decimal64(12, 0) NOT NULL COMMENT "",
  `c_1_4` int(11) NULL COMMENT "",
  `c_1_5` varchar(1) NULL COMMENT "",
  `c_1_6` decimal128(23, 10) NULL COMMENT "",
  `c_1_7` decimal64(14, 8) NULL COMMENT "",
  `c_1_8` smallint(6) NOT NULL COMMENT "",
  `c_1_9` ARRAY<int(11)> NULL COMMENT "",
  `c_1_10` ARRAY<int(11)> NOT NULL COMMENT "",
  `c_1_11` char(1) NOT NULL COMMENT "",
  `c_1_12` double NOT NULL COMMENT "",
  `c_1_13` datetime NOT NULL COMMENT "",
  `c_1_14` date NOT NULL COMMENT "",
  `c_1_15` smallint(6) NOT NULL COMMENT ""
) ENGINE=OLAP
DUPLICATE KEY(`c_1_0`, `c_1_1`, `c_1_2`, `c_1_3`, `c_1_4`, `c_1_5`, `c_1_6`, `c_1_7`, `c_1_8`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`c_1_8`, `c_1_4`, `c_1_5`, `c_1_0`, `c_1_2`, `c_1_3`, `c_1_7`, `c_1_6`) BUCKETS 10
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "DEFAULT"
);

select * from t1 where t1.c_1_9 between ([1,2,3]) and ([4,5,6]);
```

### Expected behavior (Required)
Report error since array does not support between operator.

### Real behavior (Required)
Crashed.

### StarRocks version (Required)
- 436cdc60

### LINKED ISSUE #4316: be crash (pipeline on and off)
### Steps to reproduce the behavior (Required)
```
DROP DATABASE IF EXISTS g3;
CREATE DATABASE g3;
USE g3;
admin set frontend config ("enable_decimal_v3" = "true");
CREATE TABLE t0 (c_0_0 DATETIME NOT NULL,c_0_1 DATE NOT NULL,c_0_2 LARGEINT NOT NULL,c_0_3 VARCHAR(1) NOT NULL,c_0_4 BOOLEAN NOT NULL,c_0_5 VARCHAR(1) NOT NULL,c_0_6 DATE NOT NULL,c_0_7 BOOLEAN NOT NULL,c_0_8 VARCHAR(1) NOT NULL,c_0_9 BIGINT NOT NULL,c_0_10 VARCHAR(21) NOT NULL,c_0_11 VARCHAR(31) NOT NULL,c_0_12 DECIMAL(9, 8)  NOT NULL,c_0_13 FLOAT NOT NULL) PRIMARY KEY (c_0_0,c_0_1,c_0_2,c_0_3,c_0_4,c_0_5,c_0_6,c_0_7,c_0_8,c_0_9) PARTITION BY RANGE(c_0_1) (START ("2000-01-01") END ("2021-12-31") EVERY (INTERVAL 91 day)) DISTRIBUTED BY HASH (c_0_7,c_0_1,c_0_8) properties("replication_num"="3") ;
CREATE TABLE IF NOT EXISTS t1 (c_1_0 DATE NULL ,c_1_1 BOOLEAN NOT NULL ,c_1_2 DATETIME NOT NULL ,c_1_3 BIGINT NULL ,c_1_4 VARCHAR(21) NOT NULL ,c_1_5 VARCHAR(1) NULL ,c_1_6 BOOLEAN NOT NULL ,c_1_7 DATETIME NOT NULL ,c_1_8 CHAR(11) NOT NULL ,c_1_9 VARCHAR(1) NULL ,c_1_10 DATETIME NULL ,c_1_11 HLL HLL_UNION NULL ,c_1_12 CHAR(1) MIN NOT NULL ,c_1_13 VARCHAR(21) REPLACE_IF_NOT_NULL NULL ,c_1_14 BOOLEAN REPLACE NULL ,c_1_15 BOOLEAN REPLACE NOT NULL ,c_1_16 CHAR(11) REPLACE NOT NULL ) AGGREGATE KEY (c_1_0,c_1_1,c_1_2,c_1_3,c_1_4,c_1_5,c_1_6,c_1_7,c_1_8,c_1_9,c_1_10) PARTITION BY RANGE(c_1_0) (START ("2000-01-01") END ("2021-12-31") EVERY (INTERVAL 349 day)) DISTRIBUTED BY HASH (c_1_0,c_1_5) properties("replication_num"="3") ;
INSERT INTO t1 VALUES ("2000-01-20", false, "2000-01-04 08:00:10", -1686213076, "", "", true, "2000-01-24 08:00:10", "Y", "", NULL, hll_hash(""), "", "", false, true, "");
INSERT INTO t0 VALUES ("2000-01-07 08:00:10", "2000-01-28", -1686213076, "", true, "", "2000-01-24", false, "", 901389758, "V", "PGGFBV", 7E-8, 0.7182217091880795), ("1999-12-31 08:00:10", "2000-01-05", 901389758, "", false, "", "2000-01-25", true, "", 396760070, "HS0jL22o", "LO", 0E-8, 0.27368457098563737);
INSERT INTO t0 VALUES ("2000-01-08 08:00:10", "2000-01-07", 396760070, "", false, "", "2000-01-21", false, "", 901389758, "", "8Jr", 0E-8, 0.30653431472370296);
INSERT INTO t1 VALUES ("2000-01-04", false, "2000-01-02 08:00:10", 1951322413, "5", "", false, "2000-01-27 08:00:10", "", "", NULL, hll_hash("6WDW"), "", "ciQXZ", false, true, "X");
INSERT INTO t1 VALUES ("2000-01-23", true, "2000-01-26 08:00:10", 396760070, "VvFo", "", true, "2000-01-15 08:00:10", "", "", "2000-01-03 08:00:10", hll_hash("Aw"), "", "QQe", false, true, "O"), ("2000-01-08", true, "2000-01-23 08:00:10", 1358370775, "V", "", false, "2000-01-22 08:00:10", "ki1H0Z", "", "2000-01-05 08:00:10", hll_hash("wxul76Nk"), "", "FIOqL", true, true, "z");
TRUNCATE TABLE t1;
DELETE FROM t0 WHERE c_0_7=false AND c_0_5="" AND c_0_3="" AND c_0_1="1970-01-22" AND c_0_6="1970-01-11" AND c_0_8="" AND c_0_9=-68747014;
INSERT INTO t1 VALUES ("2000-01-09", true, "2000-01-06 08:00:10", 443904903, "n", "", true, "2000-01-12 08:00:10", "ce", "", "2000-01-12 08:00:10", hll_hash("qJ8iW"), "", "22XA", false, false, ""), ("2000-01-26", false, "2000-01-31 08:00:10", -1346580012, "q9", "", false, "2000-01-28 08:00:10", "hS", "", "2000-01-10 08:00:10", hll_hash("G"), "", "SR69rWZk", false, true, "e4Mh"), ("2000-01-23", false, "2000-01-29 08:00:10", NULL, "VJmzxPv", "", false, "1999-12-31 08:00:10", "e", "", "2000-01-16 08:00:10", hll_hash("sN"), "", "", false, false, "");
SELECT t0.c_0_12 FROM t0 WHERE (["pkP", "Gb6hOE", "W3wwvV", "t", "dID2qkp0", "WsQoN", "", "2AM2", "xDB2Mg8W", "sk", "h0V8k5", "Z", "H"])  BETWEEN (["Fma6qvln", "VozJwT", "VCPwxA3Xl", "", "5Nz7CEz", "8Tf2JBdUc", "", ""]) AND (["", "", "sN7cuEwR", "xK", "wq0VIA", "N"])  GROUP BY t0.c_0_12 UNION SELECT t0.c_0_12 FROM t0 WHERE (NOT ((["pkP", "Gb6hOE", "W3wwvV", "t", "dID2qkp0", "WsQoN", "", "2AM2", "xDB2Mg8W", "sk", "h0V8k5", "Z", "H"])  BETWEEN (["Fma6qvln", "VozJwT", "VCPwxA3Xl", "", "5Nz7CEz", "8Tf2JBdUc", "", ""]) AND (["", "", "sN7cuEwR", "xK", "wq0VIA", "N"]) )) GROUP BY t0.c_0_12 UNION SELECT t0.c_0_12 FROM t0 WHERE (((["pkP", "Gb6hOE", "W3wwvV", "t", "dID2qkp0", "WsQoN", "", "2AM2", "xDB2Mg8W", "sk", "h0V8k5", "Z", "H"])  BETWEEN (["Fma6qvln", "VozJwT", "VCPwxA3Xl", "", "5Nz7CEz", "8Tf2JBdUc", "", ""]) AND (["", "", "sN7cuEwR", "xK", "wq0VIA", "N"]) ) IS NULL) GROUP BY t0.c_0_12;
```

```

start time: Mon Mar 21 23:01:43 CST 2022
*** Check failure stack trace: ***
    @          0x3d1cb5c  google::LogMessage::Fail()
    @          0x3d1cabb  google::LogMessage::SendToLog()
    @          0x3d1c484  google::LogMessage::Flush()
    @          0x3d1f750  google::LogMessageFatal::~LogMessageFatal()
    @          0x2c05e35  starrocks::type_dispatch_predicate<>()
    @          0x2c00903  starrocks::vectorized::VectorizedBinaryPredicateFactory::from_thrift()
    @          0x2b7b8b8  starrocks::Expr::create_vectorized_expr()
    @          0x2b7bde4  starrocks::Expr::create_tree_from_thrift()
    @          0x2b7be6a  starrocks::Expr::create_tree_from_thrift()
    @          0x2b7be6a  starrocks::Expr::create_tree_from_thrift()
    @          0x2b7be6a  starrocks::Expr::create_tree_from_thrift()
    @          0x2b7be6a  starrocks::Expr::create_tree_from_thrift()
    @          0x2b7c06f  starrocks::Expr::create_expr_tree()
    @          0x264a3b8  starrocks::vectorized::ProjectNode::init()
    @          0x2393a1a  starrocks::ExecNode::create_tree_helper()
    @          0x23939a9  starrocks::ExecNode::create_tree_helper()
    @          0x2393bed  starrocks::ExecNode::create_tree()
    @          0x267c7ef  starrocks::pipeline::FragmentExecutor::prepare()
    @          0x1f5d9ec  starrocks::PInternalServiceImpl<>::_exec_plan_fragment()
    @          0x1f5eff3  starrocks::PInternalServiceImpl<>::exec_plan_fragment()
    @          0x3e61f8e  brpc::policy::ProcessRpcRequest()
    @          0x3e589f7  brpc::ProcessInputMessage()
    @          0x3e598a3  brpc::InputMessenger::OnNewMessages()
    @          0x3f0053e  brpc::Socket::ProcessEvent()
    @          0x3e0e4ff  bthread::TaskGroup::task_runner()
    @          0x3f96cb1  bthread_make_fcontext
```

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
 - branch-2.2 Release
 - mysql> select current_version();
+---------------------------------+
| current_version()               |
+---------------------------------+
| SQLANCER_BRANCH_RELEASE c16172d |
+---------------------------------+

### PR DESCRIPTION
**Pull request by liuyehcf**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #4305
Fixes #4316

## Problem Summary(Required) ：

Array type couldn't as Predicate

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/catalog/Type.java [hunk 1]
```diff
@@ -582,9 +582,10 @@ public boolean isStringType() {
     // 2. don't support filter
     // 3. don't support group by
     // 4. don't support index
+    // 5. don't support array
     public boolean isOnlyMetricType() {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Type.java:586]: shouldn't change here. array could be key column， we also could group by array

**Comment by Seaven** [comment]
Need check json type? @mofeiatwork

File: fe/fe-core/src/main/java/com/starrocks/catalog/Type.java (hunk 1 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/catalog/Type.java]: Thanks, updated

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java [hunk 2]
```diff
@@ -471,8 +471,8 @@ private void predicateBaseAndCheck(Predicate node) {
             node.setType(Type.BOOLEAN);
 
             for (Expr expr : node.getChildren()) {
-                if (expr.getType().isOnlyMetricType()) {
-                    throw new SemanticException("HLL, BITMAP and PERCENTILE type couldn't as Predicate");
+                if (expr.getType().isOnlyMetricType() || expr.getType() instanceof ArrayType) {
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java:474]: How about move this logical to `Type.java` ?

**Comment by liuyehcf** [comment]
run starrocks_fe_unittest

**Comment by mofeiatwork** [comment]
> 

JSON type support predicate. These information related with specific type should be maintained at `Type.java`, which could be more clearly.

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java (hunk 2 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java:474]: ok

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 5 / 5 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/Predicate.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/catalog/Type.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|2|2|100.00%|

**Review by murphyatwork — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**